File: ios/Stockbook/DesignSystem/Components/BottomSheet.swift
import SwiftUI

/// Bottom sheets, drawn by the app rather than by `.sheet`.
///
/// The design is specific in ways the system sheet does not expose: an
/// `rgba(16,17,28,0.74)` scrim, 18px rounding on the **top corners only**, a
/// `0 -16px 40px rgba(0,0,0,0.65)` shadow, and a max height of 84%. Presenting
/// it as an overlay inside the app's own root stack gives us all of that and
/// keeps the tab bar visible behind the scrim, which is what the prototype does.
///
/// There is **no grab handle**. The design called for one and it was drawn, but
/// this sheet has no drag gesture — the handle was an invitation to do something
/// that did nothing. A sheet closes by its own Close or Done button, or by
/// tapping the scrim. Either add the gesture or do not draw the affordance;
/// drawing it alone is the one option that teaches the owner the app ignores
/// them.

/// The rounded card itself. The scrim behind it is a **sibling**, not a child —
/// see `nocturneSheet`, which is the only thing that builds either.
struct BottomSheetPanel<Content: View>: View {
    @ViewBuilder var content: Content

    var body: some View {
        // The reader wraps the scroll view because it has to be its ancestor, and
        // the proxy goes into the environment because the sheet's own content is
        // its descendant — see `sheetScroll`. Without both, a sheet can neither
        // reach its own scroll view nor scroll a field out from under the
        // keyboard.
        ScrollViewReader { proxy in
            ScrollView {
                content
                    .padding(.horizontal, Metrics.screenPadding)
                    // The header still needs air above it, or it sits against
                    // the rounded corner.
                    .padding(.top, 16)
                    .padding(.bottom, 32)
                    .environment(\.sheetScroll, proxy)
            }
            .scrollBounceBehavior(.basedOnSize)
            // The sheet is inside an overlay that ignores the keyboard, so it
            // stays put rather than leaping up — which also means nothing insets
            // its content for the keyboard unless this does.
            .keyboardRoom()
            .scrollDismissesKeyboard(.interactively)
        }
        .frame(maxWidth: .infinity)
        .background(Nocturne.surface)
        .clipShape(TopRoundedRectangle(radius: Metrics.sheetRadius))
        .shadow(color: Nocturne.sheetShadow, radius: 20, x: 0, y: -16)
    }
}

/// Rounds the top two corners only — bottom sheets sit flush against the
/// bottom edge.
struct TopRoundedRectangle: Shape {
    let radius: CGFloat

    func path(in rect: CGRect) -> Path {
        Path(
            roundedRect: rect,
            cornerRadii: RectangleCornerRadii(topLeading: radius, bottomLeading: 0, bottomTrailing: 0, topTrailing: radius),
            style: .continuous
        )
    }
}

/// The title row every sheet opens with: heading, optional sub-line, close `x`.
struct SheetHeader: View {
    let title: String
    var subtitle: String?
    let onClose: () -> Void

    var body: some View {
        HStack(alignment: .top) {
            VStack(alignment: .leading, spacing: 3) {
                Text(title).nocturneText(.sheetTitle)
                if let subtitle {
                    Text(subtitle).nocturneText(.meta)
                }
            }
            Spacer(minLength: 12)
            Button(action: onClose) {
                Glyph(Icon.close, size: 16)
                    .foregroundStyle(Nocturne.neutral500)
                    .minimumTouchTarget()
            }
            .buttonStyle(.plain)
        }
        .padding(.bottom, 14)
    }
}

extension View {
    /// Presents `content` as a Nocturne bottom sheet whenever `item` is non-nil.
    ///
    /// The scrim and the panel are separate `if`s inside one ZStack that is
    /// always mounted, and that is the whole trick. They used to be children of
    /// a single container view, each carrying its own `.transition` — and every
    /// sheet in the app faded in anyway. A `.transition` only runs when *that*
    /// view is the one being inserted; when an ancestor is inserted instead,
    /// SwiftUI animates the ancestor with the default `.opacity` and the
    /// transitions written inside it never get a turn. Lifting both out to
    /// where the condition lives is what lets the scrim fade while the panel
    /// slides up from the bottom edge.
    func nocturneSheet<Item: Identifiable, SheetContent: View>(
        item: Binding<Item?>,
        @ViewBuilder content: @escaping (Item) -> SheetContent
    ) -> some View {
        overlay {
            ZStack(alignment: .bottom) {
                if item.wrappedValue != nil {
                    SheetScrim { item.wrappedValue = nil }
                }
                if let value = item.wrappedValue {
                    BottomSheetPanel { content(value) }
                        .transition(.move(edge: .bottom))
                }
            }
            // No frame on the stack itself: while nothing is presented it holds
            // nothing, and a full-size empty layer over every screen is a thing
            // to be sure about rather than to leave lying there. Presented, the
            // scrim is a `Color` that takes whatever the overlay proposes — the
            // whole screen — and the panel settles against its bottom edge.
            .ignoresSafeArea(.keyboard, edges: .bottom)
            .animation(Metrics.sheet, value: item.wrappedValue != nil)
        }
    }

    /// The same sheet, for something that is either showing or not.
    ///
    /// Most sheets here carry the thing they are about, so `item:` is the usual
    /// form. A sheet whose whole content is a question — "which product arrived?"
    /// — has nothing to carry, and inventing an `Identifiable` box for a boolean
    /// would be ceremony.
    func nocturneSheet<SheetContent: View>(
        isPresented: Binding<Bool>,
        @ViewBuilder content: @escaping () -> SheetContent
    ) -> some View {
        overlay {
            ZStack(alignment: .bottom) {
                if isPresented.wrappedValue {
                    SheetScrim { isPresented.wrappedValue = false }
                    BottomSheetPanel { content() }
                        .transition(.move(edge: .bottom))
                }
            }
            // No frame on the stack itself: while nothing is presented it holds
            // nothing, and a full-size empty layer over every screen is a thing
            // to be sure about rather than to leave lying there. Presented, the
            // scrim is a `Color` that takes whatever the overlay proposes — the
            // whole screen — and the panel settles against its bottom edge.
            .ignoresSafeArea(.keyboard, edges: .bottom)
            .animation(Metrics.sheet, value: isPresented.wrappedValue)
        }
    }
}

/// The dimmed ground behind a sheet, and the second way out of one.
///
/// Carries its own `.transition` because it is inserted directly into the
/// sheet layer's stack, alongside the panel rather than inside it.
struct SheetScrim: View {
    let onTap: () -> Void

    var body: some View {
        Nocturne.scrim
            .ignoresSafeArea()
            .contentShape(Rectangle())
            .onTapGesture(perform: onTap)
            .transition(.opacity)
    }
}

File: android/app/src/main/kotlin/com/stockbook/app/design/BottomSheet.kt
package com.stockbook.app.design

import androidx.activity.compose.BackHandler
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp

/**
 * Bottom sheets, drawn by the app rather than by Material's `ModalBottomSheet`.
 *
 * The design is specific in ways the system sheet does not expose: an
 * `rgba(16,17,28,0.74)` scrim, rounding on the **top corners only**, a maximum
 * height of 84%, and the tab bar staying visible behind the scrim. Drawing it
 * here gives all of that.
 *
 * There is **no grab handle**. The design called for one and it was drawn, but
 * this sheet has no drag gesture — the handle was an invitation to do something
 * that did nothing. A sheet closes by its own Close or Done button, or by
 * tapping the scrim. Either add the gesture or do not draw the affordance;
 * drawing it alone is the one option that teaches the owner the app ignores
 * them.
 */
@Composable
fun BottomSheet(
    visible: Boolean,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit
) {
    if (!visible) return

    // The system back gesture closes the sheet, not the app.
    //
    // Written once here rather than at each of the dozen call sites, and it lands
    // on the right one for free: `BackHandler` hands the press to the callback
    // registered *last*, and a sheet is composed after the screen it covers. A
    // sheet opened over the statement therefore closes itself and leaves the
    // statement standing, which is what the layering already says on screen.
    BackHandler(onBack = onDismiss)

    Box(modifier = modifier.fillMaxSize()) {
        // The scrim swallows taps, which is how a sheet is dismissed.
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Nocturne.scrim)
                .clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                    onClick = onDismiss
                )
        )

        AnimatedVisibility(
            visible = true,
            enter = slideInVertically { it } + fadeIn(),
            exit = slideOutVertically { it } + fadeOut(),
            modifier = Modifier.align(Alignment.BottomCenter)
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier
                    .fillMaxWidth()
                    .fillMaxHeight(0.84f)
                    .clip(RoundedCornerShape(topStart = Metrics.sheetRadius, topEnd = Metrics.sheetRadius))
                    .background(Nocturne.surface)
                    // A tap inside must not reach the scrim behind it.
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        onClick = {}
                    )
            ) {
                // The header still needs air above it, or it sits against the
                // rounded corner.
                Spacer(Modifier.height(16.dp))

                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .verticalScroll(rememberScrollState())
                        .padding(horizontal = Metrics.screenPadding)
                        .navigationBarsPadding()
                        // Every sheet in the app carries fields, so this one line
                        // lifts all of them clear of the keyboard rather than each
                        // sheet remembering to.
                        .imePadding()
                        .padding(bottom = 32.dp)
                ) {
                    content()
                }
            }
        }
    }
}

/** The title row every sheet opens with: heading, optional sub-line, close. */
@Composable
fun SheetHeader(
    title: String,
    onClose: () -> Unit,
    modifier: Modifier = Modifier,
    subtitle: String? = null
) {
    androidx.compose.foundation.layout.Row(
        verticalAlignment = Alignment.Top,
        modifier = modifier.fillMaxWidth().padding(bottom = 14.dp)
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(title, style = NocturneType.sheetTitle, color = Nocturne.text)
            if (subtitle != null) {
                Text(
                    subtitle,
                    style = NocturneType.meta,
                    color = Nocturne.neutral500,
                    modifier = Modifier.padding(top = 3.dp)
                )
            }
        }
        IconButton(Icon.close, onClick = onClose, size = 16.dp, tint = Nocturne.neutral500)
    }
}
